### DISCUSSION TITLE
Custom directives

### DISCUSSION THREAD

**Discussion by jaroslaw-blazkow** [Netflix/dgs-framework#69] [Q&A]
Hi, I'm trying to write some custom directives using DGS.
As i didn't find anything regarding this topic in  DGS docu, I tried to implement it like graphql-java suggests.
Create custom runtimeWiring bean with added class which implements SchemaDirectiveWiring. 
But this wiring is not taken account. Any advices how it should be approached in DGS?

**Comment by jaroslaw-blazkow** [comment]
I found solution partly based on another post:

@DgsComponent
public class UppercaseDirective implements SchemaDirectiveWiring
{
  @DgsRuntimeWiring
  public RuntimeWiring.Builder runtimeWiring( RuntimeWiring.Builder builder )
  {
    return builder.directiveWiring( new UppercaseDirective() );
  }
  
  @Override
  public GraphQLArgument onArgument( SchemaDirectiveWiringEnvironment<GraphQLArgument> environment )
  {
...
}

**Reply by golfingal72** [reply]
@BurnedEyes 

I've tried adding a new schema directive wiring as you indicate and can see it is getting wired at startup, but I can't get it to execute..  is there something I'm missing?  do you have a full example in the Dgs environment?

**Reply by jaroslaw-blazkow** [reply]
```
directive @stringCase(case: Case) on FIELD_DEFINITION
 
 
enum Case {
  UPPER, LOWER
}
```

```@DgsComponent
public class StringCaseDirective implements SchemaDirectiveWiring
{
  @DgsRuntimeWiring
  public RuntimeWiring.Builder runtimeWiring( RuntimeWiring.Builder builder )
  {
    return builder.directiveWiring( new StringCaseDirective() );
  }
 
  @Override
  public GraphQLFieldDefinition onField( SchemaDirectiveWiringEnvironment<GraphQLFieldDefinition> env )
  {
    GraphQLFieldDefinition field = env.getElement();
    GraphQLFieldsContainer parentType = env.getFieldsContainer();
    if (field.getDirective("stringCase") != null) {
      GraphQLArgument argument = field.getDirective("stringCase").getArgument("case");
      String argumentValue = (String) argument.getValue();
 
      // build a data fetcher that transforms the given value to uppercase/lowercase
      DataFetcher originalFetcher = env.getCodeRegistry().getDataFetcher(parentType, field);
      DataFetcher dataFetcher = DataFetcherFactories.wrapDataFetcher(originalFetcher, ((dataFetchingEnvironment, value) -> {
        if (value instanceof String) {
          if (argumentValue.equalsIgnoreCase("UPPER")) {
            return ((String) value).toUpperCase();
          }
          if (argumentValue.equalsIgnoreCase("LOWER")) {
            return ((String) value).toLowerCase();
          }
        }
        return value;
      }));
 
      // now change the field definition to use the new uppercase/lowercase data fetcher
      env.getCodeRegistry().dataFetcher(parentType, field, dataFetcher);
    }
    return field;
  }
 
}```

**Reply by golfingal72** [reply]
Thanks so much... really helpful!

**Comment by paulbakker** [comment]
It's actually simpler than that!
You can add a directive to the schema, you don't need to "register" it to the framework.

```graphqls
hello: String  @mydirective(message: "hello")

directive @mydirective(message: String) on FIELD_DEFINITION
```

```java
@DgsData(parentType = "query", field = "hello")
public String hello(DataFetchingEnvironment dfe) {
   GraphQLDirective mydirective = dfe.getFieldDefinition().getDirective("mydirective");
   String message = mydirective.getArgument("message").getValue();
   return message;
}
```

**Reply by jaroslaw-blazkow** [reply]
thanks, but looks like there's no common logic in that implementation for custom directive

**Reply by paulbakker** [reply]
What do you mean by common logic? Can you give an example?

**Reply by jregistr** [reply]
@BurnedEyes Did you want the presence of the directive to run some kind of function before and/or after the annotated field's resolver runs?

**Reply by jaroslaw-blazkow** [reply]
@jregistr - exactly, I just respond to @paulbakker simpler solution - which is fine, but doesn't allow you to specify common logic for directive used in many places.

**Reply by golfingal72** [reply]
common logic is to uppercase fields at will in the original example, 
the way you coded it, your individual resolvers would have to know the directive was active... that's alot of custom code
if the directive was used in many places

**Reply by WtfJoke** [reply]
I figured out that the provided example only works with directives on field_definition. 

Is there such a simple solution for directives based on `INPUT_FIELD_DEFINITION` or `ARGUMENT_DEFINITION`?

**Comment by WtfJoke** [comment]
I also created a sample repo to demonstrate using directives (provided by [graphql-java/graphql-java-extended-validation](https://github.com/graphql-java/graphql-java-extended-validation)) of and wire it using DGS. Feel free to check it out and make comments:


https://github.com/WtfJoke/graphql-dgs-directives-demo
